spec/integ/Ipv6Generator.spec.js

```diff
@@ -0,0 +1,145 @@
+const { IrcServer } = require("../../lib/irc/IrcServer");
+
+describe("Ipv6Generator", () => {
+    const {env, roomMapping, test} = envBundle();
+    let generator;
+    let store;
+    const server = new IrcServer("domain", IrcServer.DEFAULT_CONFIG, "example.com");
+
+    beforeEach(async () => {
+        await test.beforeEach(env);
+        env.ircMock._whenClient(roomMapping.server, roomMapping.botNick, "connect", (client, fn) => {
+            fn();
+        });
+        await test.initEnv(env);
+        store = env.ircBridge.getStore();
+        generator = new Ipv6Generator(store);
+    });
+
+    afterEach(async () => test.afterEach(env));
+
+
+    it("should generate an IPv6 address", async function() {
+        const address = await generator.generate(
+            '2001:0db8:85a3::', new IrcClientConfig('@foo:example.com', 'irc.example.com'), server
+        );
+        expect(address).toEqual('2001:0db8:85a3::1');
+        const newConfig = await store.getIrcClientConfig('@foo:example.com', 'irc.example.com');
+        expect(newConfig.userId).toEqual('@foo:example.com');
+        expect(newConfig.domain).toEqual('irc.example.com');
+        expect(newConfig.config.ipv6).toEqual('2001:0db8:85a3::1');
+    });
+
+    it("should NOT generate an IPv6 address for an existing config", async function() {
+        const config = new IrcClientConfig('@foo:example.com', 'irc.example.com', {
+            ipv6: '2001:0db8:85a3::1a16'
+        });
+        await store.storeIrcClientConfig(config);
+        const address = await generator.generate(
+            '2001:0db8:85a3::', new IrcClientConfig('@foo:example.com', 'irc.example.com',
+                {ipv6: '2001:0db8:85a3::1a16'}
+            ), server);
+        expect(address).toEqual('2001:0db8:85a3::1a16');
+        const fetchedConfig = await store.getIrcClientConfig('@foo:example.com', 'irc.example.com');
+        expect(fetchedConfig.userId).toEqual('@foo:example.com');
+        expect(fetchedConfig.domain).toEqual('irc.example.com');
+        expect(fetchedConfig.config.ipv6).toEqual('2001:0db8:85a3::1a16');
+    });
+
+    it("should queue and generate multiple IPv6 addresses", async function() {
+        const promises = [];
+        for (let i = 0; i < 10; i++) {
+            promises.push(
+                generator.generate(
+                    '2001:0db8:85a3::',
+                    new IrcClientConfig(`@foo${i}:example.com`, 'irc.example.com'),
+                    server,
+                )
+            );
+        }
+        await Promise.all(promises);
+        for (let i = 0; i < 10; i++) {
+            const newConfig = await store.getIrcClientConfig(`@foo${i}:example.com`, 'irc.example.com');
+            expect(newConfig.userId).toEqual(`@foo${i}:example.com`);
+            expect(newConfig.domain).toEqual('irc.example.com');
+            expect(newConfig.config.ipv6).toEqual(`2001:0db8:85a3::${(i+1).toString(16)}`);
```

Comment:
Can we – and do we want to – guarantee that they'll come out in this specific order?

It may be safer to just check if each userId and IP show up, but not necessarily in the order we requested them – unless there's a good reason to guarantee it?

Reply:
I think it's to make finding a free connection easier, if we count up from 0 then we always know if there is a free one if we store a counter. If we randomise the IPs, then we have to do extra DB calls. 

I see what you mean in that the tests don't need to check for this, but I think it's okay to be strict in this case given our code is strictly ordered.

Reply:
Make sense to me :+1: